File: Python/304. Range Sum Query 2D - Immutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        self.mat_sum = [[0 for i in range(len(matrix[0]))] for j in range(len(matrix))]
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if i > 0:
                    self.mat_sum[i][j] += self.mat_sum[i - 1][j]
                if j > 0:
                    self.mat_sum[i][j] += self.mat_sum[i][j - 1]
                if i > 0 and j > 0:
                    self.mat_sum[i][j] -= self.mat_sum[i - 1][j - 1]
                self.mat_sum[i][j] += matrix[i][j]        

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        ans = self.mat_sum[row2][col2]        
        if row1 > 0:
            ans -= self.mat_sum[row1 - 1][col2]            
        if col1 > 0:
            ans -= self.mat_sum[row2][col1 - 1]            
        if row1 > 0 and col1 > 0:
            ans += self.mat_sum[row1 - 1][col1 - 1]                    
        return ans


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)

File: 2021/304. Range Sum Query 2D - Immutable.cpp
class NumMatrix {
public:
    vector<vector<int>> preSum;
    NumMatrix(vector<vector<int>>& matrix) {
        preSum.resize(matrix.size(), vector<int>(matrix[0].size(), 0));
        for (int i = 0; i < matrix.size(); i++) {            
            for (int j = 0; j < matrix[0].size(); j++) {
                preSum[i][j] = matrix[i][j];
                if (i > 0) {
                    preSum[i][j] += preSum[i - 1][j];
                }
                if (j > 0) {
                    preSum[i][j] += preSum[i][j - 1];
                }
                if (i > 0 && j > 0) {
                    preSum[i][j] -= preSum[i - 1][j - 1];
                }         
            }            
        }        
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        int sum = preSum[row2][col2];
        if (row1 > 0) {
            sum -= preSum[row1 - 1][col2];
        }
        if (col1 > 0) {
            sum -= preSum[row2][col1 - 1];
        }
        if (row1 > 0 && col1 > 0) {
            sum += preSum[row1 - 1][col1 - 1];
        }
        return sum;
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */

 class NumMatrix {
public:
    vector<vector<int>> preSum;
    NumMatrix(vector<vector<int>>& mat) {
         int m = mat.size(), n = mat[0].size();
        preSum.resize(m + 1, vector<int>(n + 1, 0));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                preSum[i + 1][j + 1] = mat[i][j] + preSum[i][j + 1] + preSum[i + 1][j] - preSum[i][j];
            }
        }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return preSum[row2 + 1][col2 + 1] - preSum[row1][col2 + 1] - preSum[row2 + 1][col1] + preSum[row1][col1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */